README.md
term-ctrl
=========

[<img alt="GitHub Workflow Status" src="https://img.shields.io/github/workflow/status/jnqnfe/term-ctrl-rust/Build%20&%20Test/master?style=for-the-badge" height="24">](https://github.com/jnqnfe/term-ctrl-rust/actions)
[<img alt="crates.io" src="https://img.shields.io/crates/v/term-ctrl?style=for-the-badge" height="24">](https://crates.io/crates/lterm-ctrl)
[<img alt="docs.rs" src="https://img.shields.io/crates/v/term-ctrl?color=5479ab&label=docs.rs&style=for-the-badge" height="24">](https://docs.rs/term-ctrl)
[<img alt="min-rust-version" src="https://img.shields.io/static/v1?label=RUST&message=1.31%2B&color=informational&style=for-the-badge" height="24">](https://rust-lang.github.io/rfcs/2495-min-rust-version.html)

A small library for ANSI terminal output formatting assistance.

This is an alternative or companion to the excellent `ansi_term` crate.


CHANGELOG.md
# 0.7.8 (August 1st, 2021)

 * Fixed CI failures, finally
 * Excluded the CI config from being bundled in the crate

# 0.7.7 (August 1st, 2021)

 * Actually drop the travis config
 * Add scheduled CI run

# 0.7.6 (December 15th, 2020)

 * Moved CI to github actions

# 0.7.5 (November 25th, 2020)

 * Fixed deprecated license attribute syntax

# 0.7.4 (November 25th, 2020)

 * Another attempt to fix the macro link

# 0.7.3 (November 24th, 2020)

 * Trying again to fix mistake with macro link in previous changes (relying upon docs.rs processing
   of new crates to test the change since my distro only has rustc v1.47 for now).

# 0.7.2 (November 24th, 2020)

 * Fix mistake with macro link in previous changes

# 0.7.1 (November 24th, 2020)

 * Previous change was incomplete

# 0.7.0 (November 24th, 2020)

 * Adjusted documentation to take advantage of the better link generation provided in the latest
   rust compiler (version 1.48). This means that the minimum rust version needed is 1.48 to have
   working links within the documentation, though the crate is still usable with older versions, so
   the MSRV is not being bumped.

# 0.6.1 (November 24th, 2020)

 * Trivial non-functional tweaks, not worth listing

# 0.6.0 (September 17th, 2019)

 * Re-organised the predefines
 * Added various additional predefines
 * Renamed the `REVERSE` predefine to the more correct term `INVERSE`
 * Extended, corrected and otherwise further improved the documentation
 * Added helper macros for generating code groupings for use in sequences, including the generic
   `codes`, along with those specifically for specifying 256-colour and RGB colour palette
   selections: `c256_fg`, `c256_bg`, `rgb_fg` and `rgb_bg`.
 * Added a `codes` mod. While the `predefines` mod offers various common items in full sequence
   form, this new mod offers the codes without being embedded in the full sequence pattern. This may
   be useful to some for constructing custom combinations via named constants instead of raw numeric
   literals. The prefix and postfix parts of the sequence pattern are given also. Note that it is
   not possible to use these names constants with `seq` since the Rust macro construction would take
   the literal name of the constant, rather than what the constant points to.
 * Added a `demo` example, which outputs some sample text with various forms of formatting applied,
   to demonstrate the possibilities and the capabilities of your terminal.

# 0.5.0 (September 15th, 2019)

 * Reworked the format-supported helpers on top of the `atty` crate, thus bringing better cross
   platform support.
 * Added a Windows `enable_ansi_support` function, taken from `ansi_term`
 * Renamed the `term_seq` macro to more simply `seq`
 * Reorganised a bit
 * Re-wrote the documentation
 * Various other odd little tweaks really not worth documenting

# 0.4.1 (October 27th, 2018)

 * Documentation example tweaks

# 0.4.0 (October 16th, 2018)

 * Introduced unit tests
 * Used a trait to help ensure that the `fmt_supported` method on `StdPipe` always has the same
   prototype across platforms.
 * Removed a stray redundant platform conditional
 * Enabled some doc tests
 * Reorganised copyright header position

# 0.3 (October 10th, 2018)

 * Grouped predefined sequences in `predefined` mod

# 0.2 (October 9th, 2018)

 * Improved crate documentation
 * Now avoids `libc` for windows builds, where it was unused
 * Grouped codes into sub-modules
 * Changed codes from statics to constants
 * Added and used a macro for creating sequences
 * Renamed `color_supported` to `fmt_supported`
 * Expanded pipe type coverage
 * Declared `no_std` since we do not need `std`

# 0.1 (January 24th, 2018)

 * Original version, previously bundled with the `gong` crate & repo, now extracted into its own


tests.rs
// Copyright 2017 Lyndon Brown
//
// Licensed under the MIT license or the Apache license (version 2.0), at your option. You may not
// copy, modify, or distribute this file except in compliance with said license. You can find copies
// of these licenses either in the LICENSE-MIT and LICENSE-APACHE files, or alternatively at
// <http://opensource.org/licenses/MIT> and <http://www.apache.org/licenses/LICENSE-2.0>
// respectively.

//! Unit tests

#[macro_use]
extern crate term_ctrl;

use term_ctrl::*;

/// Check `seq` macro is available and working correctly
#[test]
fn term_seq() {
    assert_eq!(seq!(1), "\u{1B}[1m");
    assert_eq!(seq!(1,2), "\u{1B}[1;2m");
    assert_eq!(seq!(4,8,15,16,23,42), "\u{1B}[4;8;15;16;23;42m");
}

/// Check a few `predefined` sequences seem to be available and as expected
#[test]
fn predefines() {
    assert_eq!(predefined::RESET, "\u{1B}[0m");
    assert_eq!(predefined::effects::BOLD, "\u{1B}[1m");
    assert_eq!(predefined::color1::GREEN, "\u{1B}[32m");
    assert_eq!(predefined::color1_bold::GREEN, "\u{1B}[32;1m");
}

#[cfg(not(windows))]
mod platform {
    use super::*;

    /* On non-Windows:
     * - It should always return false for `StdPipe::StdIn`
     * - Stdout/stderr have dynamic support however, depending on whether connected to a tty
     *
     * How to handle this dynamic situation?
     *  1. We could not bother to compare the returned value, just that the functions are reachable?
     *  2. We could assume that stdout/stderr will always be a tty in running tests?
     *  3. We could try to be fancy and set up both types of pipes, if possible?
     */
    //TODO: Consider switching to option #3?

    #[test]
    fn fmt_supported_shortcuts() {
        // Dynamic, assuming test env has them connected to tty
        assert_eq!(true, fmt_supported_stdout());
        assert_eq!(true, fmt_supported_stderr());
    }

    #[test]
    fn fmt_supported() {
        // Note, stdin obviously never supports formatted output!
        assert_eq!(false, StdPipe::fmt_supported(StdPipe::StdIn));
        // Dynamic, assuming test env has them connected to tty
        assert_eq!(true, StdPipe::fmt_supported(StdPipe::StdOut));
        assert_eq!(true, StdPipe::fmt_supported(StdPipe::StdErr));
    }

    #[test]
    fn fmt_supported_withpref() {
        // Dynamic, assuming test env has them connected to tty
        assert_eq!(true, use_fmt_stdout(true));
        assert_eq!(true, use_fmt_stderr(true));
        assert_eq!(false, use_fmt_stdout(false));
        assert_eq!(false, use_fmt_stderr(false));
    }
}

#[cfg(windows)]
mod platform {
    use super::*;

    #[test]
    fn fmt_supported_shortcuts() {
        assert_eq!(false, fmt_supported_stdout());
        assert_eq!(false, fmt_supported_stderr());
    }

    #[test]
    fn fmt_supported() {
        assert_eq!(false, StdPipe::fmt_supported(StdPipe::StdIn));
        assert_eq!(false, StdPipe::fmt_supported(StdPipe::StdOut));
        assert_eq!(false, StdPipe::fmt_supported(StdPipe::StdErr));
    }

    #[test]
    fn fmt_supported_withpref() {
        assert_eq!(false, use_fmt_stdout(true));
        assert_eq!(false, use_fmt_stderr(true));
        assert_eq!(false, use_fmt_stdout(false));
        assert_eq!(false, use_fmt_stderr(false));
    }
}


tests/tests.rs
// Copyright 2017 Lyndon Brown
//
// Licensed under the MIT license or the Apache license (version 2.0), at your option. You may not
// copy, modify, or distribute this file except in compliance with said license. You can find copies
// of these licenses either in the LICENSE-MIT and LICENSE-APACHE files, or alternatively at
// <http://opensource.org/licenses/MIT> and <http://www.apache.org/licenses/LICENSE-2.0>
// respectively.

//! Unit tests

extern crate term_ctrl;

use term_ctrl::*;

/// Check `seq` macro is available and working correctly
#[test]
fn term_seq() {
    assert_eq!(seq!(1), "\u{1B}[1m");
    assert_eq!(seq!(1,2), "\u{1B}[1;2m");
    assert_eq!(seq!(4,8,15,16,23,42), "\u{1B}[4;8;15;16;23;42m");
}

/// Check a few `predefined` sequences seem to be available and as expected
#[test]
fn predefines() {
    assert_eq!(predefined::RESET, "\u{1B}[0m");
    assert_eq!(predefined::effects::BOLD, "\u{1B}[1m");
    assert_eq!(predefined::colours::fg::GREEN, "\u{1B}[32m");
    assert_eq!(predefined::combinations::fg_bold::GREEN, "\u{1B}[32;1m");
}

#[cfg(not(windows))]
mod platform {
    use super::*;

    #[test]
    fn fmt_supported() {
        // Assuming test env has them connected to tty
        assert_eq!(true, support::fmt_supported_stdout());
        assert_eq!(true, support::fmt_supported_stderr());
    }

    #[test]
    fn fmt_supported_withpref() {
        // Assuming test env has them connected to tty
        assert_eq!(true, support::use_fmt_stdout(true));
        assert_eq!(true, support::use_fmt_stderr(true));
        assert_eq!(false, support::use_fmt_stdout(false));
        assert_eq!(false, support::use_fmt_stderr(false));
    }
}


src/codes.rs
// Copyright 2017 Lyndon Brown
//
// Licensed under the MIT license or the Apache license (version 2.0), at your option. You may not
// copy, modify, or distribute this file except in compliance with said license. You can find copies
// of these licenses either in the LICENSE-MIT and LICENSE-APACHE files, or alternatively at
// <http://opensource.org/licenses/MIT> and <http://www.apache.org/licenses/LICENSE-2.0>
// respectively.

//! Code constants
//!
//! In string form.
//!
//! If you want whole predefined sequences, see the `predefined` mod. This mode offers simply the
//! numeric codes (as strings), should that be useful to anyone. Note that you cannot give these to
//! the `seq` macro; you will just get an error unfortunately - a limitation of the macro system.

use crate::codes;

// Alias for the non-British-English speakers
pub use self::colours as colors;

/// The start of a control sequence
pub const SEQ_PREFIX: &str = "\u{1B}[";
/// The end of a control sequence
pub const SEQ_POSTFIX: &str = "m";

/// Resets everything to defaults (removes all effects and colours specified)
pub const RESET: &str = codes!(0);

/// Effects
pub mod effects {
    use super::codes;

    /// Alias for reset-all
    pub const NORMAL:      &str = super::RESET;
    /// Bold (increase intensity)
    pub const BOLD:        &str = codes!(1);
    /// Dim (aka faint; decrease intensity)
    pub const DIM:         &str = codes!(2);
    /// Italic
    pub const ITALIC:      &str = codes!(3);
    /// Underline
    pub const UNDERLINE:   &str = codes!(4);
    /// Blink
    pub const BLINK:       &str = codes!(5);
    /// Rapid-blink
    pub const RAPID_BLINK: &str = codes!(6);
    /// Inverse (swap foreground/background colours)
    pub const INVERSE:     &str = codes!(7);
    /// Invisible (hidden)
    pub const INVISIBLE:   &str = codes!(8);
    /// Strike-through
    pub const STRIKE:      &str = codes!(9);

    /* `10`-`19` are used for font selection, covered by the font mod */

    /// Fraktur
    pub const FRAKTUR:       &str = codes!(20);
    /// Double-underline
    pub const DBL_UNDERLINE: &str = codes!(21);

    /// Alias for removing blink
    pub const STEADY:    &str = remove::BLINK;
    /// Alias for removing inverse
    pub const POSITIVE:  &str = remove::INVISIBLE;
    /// Alias for v invisible
    pub const VISIBLE:   &str = remove::INVISIBLE;

    /// Sequences that remove specific effects
    pub mod remove {
        use super::codes;

        /// Removes bold and/or dim
        pub const BOLD_DIM:  &str = codes!(22);
        /// Removes italic
        pub const ITALIC:    &str = codes!(23);
        /// Removes underline
        pub const UNDERLINE: &str = codes!(24);
        /// Removes blink
        pub const BLINK:     &str = codes!(25);
        /* `26` is unused? */
        /// Removes inverse
        pub const INVERSE:   &str = codes!(27);
        /// Removes invisible
        pub const INVISIBLE: &str = codes!(28);
        /// Removes strike-through
        pub const STRIKE:    &str = codes!(29);

        /// Alias for `BOLD_DIM` (reset intensity)
        pub const INTENSITY:  &str = BOLD_DIM;
    }
}

/// Font selection
pub mod fonts {
    use super::codes;

    /// Select the primary (default) font
    pub const DEFAULT: &str = codes!(10);
    /// Select alternate font #1
    pub const ALT1: &str = codes!(11);
    /// Select alternate font #2
    pub const ALT2: &str = codes!(12);
    /// Select alternate font #3
    pub const ALT3: &str = codes!(13);
    /// Select alternate font #4
    pub const ALT4: &str = codes!(14);
    /// Select alternate font #5
    pub const ALT5: &str = codes!(15);
    /// Select alternate font #6
    pub const ALT6: &str = codes!(16);
    /// Select alternate font #7
    pub const ALT7: &str = codes!(17);
    /// Select alternate font #8
    pub const ALT8: &str = codes!(18);
    /// Select alternate font #9
    pub const ALT9: &str = codes!(19);
}

/// Text foreground and background colours
pub mod colours {
    use super::codes;

    /// Reset both foreground and background colours
    pub const RESET:    &str = codes!(39, 49);
    /// Alias for resetting foreground colour
    pub const RESET_FG: &str = fg::RESET;
    /// Alias for resetting background colour
    pub const RESET_BG: &str = bg::RESET;

    /// Text (foreground) colour
    pub mod fg {
        use super::codes;

        pub const BLACK:   &str = codes!(30);
        pub const RED:     &str = codes!(31);
        pub const GREEN:   &str = codes!(32);
        pub const YELLOW:  &str = codes!(33);
        pub const BLUE:    &str = codes!(34);
        pub const MAGENTA: &str = codes!(35);
        pub const CYAN:    &str = codes!(36);
        pub const WHITE:   &str = codes!(37);

        /// Resets foreground colour to default
        pub const RESET:   &str = codes!(39);

        /// Bright variants
        pub mod bright {
            use super::codes;

            pub const BLACK:   &str = codes!(90);
            pub const RED:     &str = codes!(91);
            pub const GREEN:   &str = codes!(92);
            pub const YELLOW:  &str = codes!(93);
            pub const BLUE:    &str = codes!(94);
            pub const MAGENTA: &str = codes!(95);
            pub const CYAN:    &str = codes!(96);
            pub const WHITE:   &str = codes!(97);
        }
    }

    /// Text (background) highlighting colour
    pub mod bg {
        use super::codes;

        pub const BLACK:   &str = codes!(40);
        pub const RED:     &str = codes!(41);
        pub const GREEN:   &str = codes!(42);
        pub const YELLOW:  &str = codes!(43);
        pub const BLUE:    &str = codes!(44);
        pub const MAGENTA: &str = codes!(45);
        pub const CYAN:    &str = codes!(46);
        pub const WHITE:   &str = codes!(47);

        /// Resets background-highlight colour to default
        pub const RESET:   &str = codes!(49);

        /// Bright variants
        pub mod bright {
            use super::codes;

            pub const BLACK:   &str = codes!(100);
            pub const RED:     &str = codes!(101);
            pub const GREEN:   &str = codes!(102);
            pub const YELLOW:  &str = codes!(103);
            pub const BLUE:    &str = codes!(104);
            pub const MAGENTA: &str = codes!(105);
            pub const CYAN:    &str = codes!(106);
            pub const WHITE:   &str = codes!(107);
        }
    }
}

/// Miscellaneous
pub mod misc {
    use super::codes;

    /// Framed
    pub const FRAMED:    &str = codes!(51);
    /// Encircled
    pub const ENCIRCLED: &str = codes!(52);
    /// Overlined
    pub const OVERLINED: &str = codes!(53);

    /// Remove misc effects
    pub mod remove {
        use super::codes;

        /// Remove framed and encircled effects
        pub const FRAMED_ENCIRCLED: &str = codes!(54);
        /// Remove overlined effects
        pub const OVERLINED:        &str = codes!(55);
    }

    /// Ideogram stuff
    pub mod ideogram {
        use super::codes;

        /// Ideogram underline or right side line
        pub const UNDERLINE:      &str = codes!(60);
        /// Ideogram double underline or double line on the right side
        pub const DBL_UNDERLINE:  &str = codes!(61);
        /// Ideogram overline or left side line
        pub const OVERLINE:       &str = codes!(62);
        /// Ideogram double overline or double line on the left side
        pub const DBL_OVERLINE:   &str = codes!(63);
        /// Ideogram stress marking
        pub const STRESS_MARKING: &str = codes!(64);
        /// Ideogram attributes off
        pub const RESET:          &str = codes!(65);
    }
}

/// Combinations
pub mod combinations {
    use super::codes;

    /// Combined bold + text (foreground) colour
    pub mod fg_bold {
        use super::*;

        pub const BLACK:   &str = codes!(30,1);
        pub const RED:     &str = codes!(31,1);
        pub const GREEN:   &str = codes!(32,1);
        pub const YELLOW:  &str = codes!(33,1);
        pub const BLUE:    &str = codes!(34,1);
        pub const MAGENTA: &str = codes!(35,1);
        pub const CYAN:    &str = codes!(36,1);
        pub const WHITE:   &str = codes!(37,1);
    }
}


src/macros.rs
// Copyright 2019 Lyndon Brown
//
// Licensed under the MIT license or the Apache license (version 2.0), at your option. You may not
// copy, modify, or distribute this file except in compliance with said license. You can find copies
// of these licenses either in the LICENSE-MIT and LICENSE-APACHE files, or alternatively at
// <http://opensource.org/licenses/MIT> and <http://www.apache.org/licenses/LICENSE-2.0>
// respectively.

//! Macros

/// Constructs a control sequence
///
/// A sequence can be created with one or more numbers separated by commas.
///
/// # Examples:
///
/// ```rust
/// # use term_ctrl::seq;
/// // Bold
/// assert_eq!("\u{1B}[1m", seq!(1));
/// // Red + bold + underline
/// assert_eq!("\u{1B}[31;1;4m", seq!(31,1,4));
/// ```
#[macro_export]
macro_rules! seq {
    ($n1:expr, $($n2:expr),*) => { concat!("\u{1B}[", $n1, $(concat!(";", $n2)),*, "m") };
    ($n1:expr, $($n2:expr,)*) => { concat!("\u{1B}[", $n1, $(concat!(";", $n2)),*, "m") };
    ($n:expr) => { concat!("\u{1B}[", $n, "m") };
    () => { "" };
}

/// Constructs a code set (to be used in a control sequence)
///
/// This simply turns a sequence of numbers into a string with a semi-colon separator.
///
/// Note that this will actually happily accept strings also, not just numeric literals.
///
/// # Examples:
///
/// ```rust
/// # use term_ctrl::{seq, codes};
/// assert_eq!("1", codes!(1));
/// assert_eq!("1;2;3", codes!(1, 2, 3));
/// assert_eq!("1;2;3;4;5", codes!(1, "2;3;4", 5));
/// // Use in a sequence
/// assert_eq!("\u{1B}[1;2;3m", seq!(codes!(1,2,3)));
/// assert_eq!("\u{1B}[1;2;3;4;5;6m", seq!(1,2,codes!(3,4,5),6));
/// ```
#[macro_export]
macro_rules! codes {
    ($n1:expr, $($n2:expr),*) => { concat!($n1, $(concat!(";", $n2)),*) };
    ($n1:expr, $($n2:expr,)*) => { concat!($n1, $(concat!(";", $n2)),*) };
    ($n:expr) => { concat!($n) };
    () => { "" };
}

/// Constructs a 256-colour foreground (text) colour code set (to be used in a control sequence)
///
/// # Examples:
///
/// ```rust
/// # use term_ctrl::{seq, c256_fg};
/// // Get the codes
/// assert_eq!("38;5;238", c256_fg!(238));
/// // Use in a sequence
/// assert_eq!("\u{1B}[38;5;238m", seq!(c256_fg!(238)));
/// ```
#[macro_export]
macro_rules! c256_fg {
    ($col:expr) => { concat!("38;5;", $col) };
}

/// Constructs a 256-colour background colour code set (to be used in a control sequence)
///
/// # Examples:
///
/// ```rust
/// # use term_ctrl::{seq, c256_bg};
/// // Get the codes
/// assert_eq!("48;5;238", c256_bg!(238));
/// // Use in a sequence
/// assert_eq!("\u{1B}[48;5;238m", seq!(c256_bg!(238)));
/// ```
#[macro_export]
macro_rules! c256_bg {
    ($col:expr) => { concat!("48;5;", $col) };
}

/// Constructs an RGB foreground (text) colour code set (to be used in a control sequence)
///
/// # Examples:
///
/// ```rust
/// # use term_ctrl::{seq, rgb_fg};
/// // Get the RGB codes
/// assert_eq!("38;2;180;15;70", rgb_fg!(180, 15, 70));
/// // Use in a sequence
/// assert_eq!("\u{1B}[38;2;180;15;70m", seq!(rgb_fg!(180, 15, 70)));
/// ```
#[macro_export]
macro_rules! rgb_fg {
    ($red:expr, $green:expr, $blue:expr) => { concat!("38;2;", $red, ";", $green, ";", $blue) };
}

/// Constructs an RGB background colour code set (to be used in a control sequence)
///
/// # Examples:
///
/// ```rust
/// # use term_ctrl::{seq, rgb_bg};
/// // Get the RGB codes
/// assert_eq!("48;2;180;15;70", rgb_bg!(180, 15, 70));
/// // Use in a sequence
/// assert_eq!("\u{1B}[48;2;180;15;70m", seq!(rgb_bg!(180, 15, 70)));
/// ```
#[macro_export]
macro_rules! rgb_bg {
    ($red:expr, $green:expr, $blue:expr) => { concat!("48;2;", $red, ";", $green, ";", $blue) };
}


src/lib.rs
// Copyright 2017 Lyndon Brown
//
// Licensed under the MIT license or the Apache license (version 2.0), at your option. You may not
// copy, modify, or distribute this file except in compliance with said license. You can find copies
// of these licenses either in the LICENSE-MIT and LICENSE-APACHE files, or alternatively at
// <http://opensource.org/licenses/MIT> and <http://www.apache.org/licenses/LICENSE-2.0>
// respectively.

//! ANSI terminal text formatting assistant
//!
//! Some virtual-terminals (ANSI ones being the focus here) have a feature of allowing command-line
//! programs executed in them to request that certain limited formatting effects such as colour or
//! boldness be applied to portions of the text the program outputs for them to display. This is
//! achieved through the program embedding special “ANSI control” byte sequences into that text
//! output. This can prove very useful for instance in highlighting error notices.
//!
//! The purpose of this crate is to offer functionality to assist in generating and using such
//! control sequences.
//!
//! Note that such formatting should only be used when the `stdout` stream (or `stderr` as
//! applicable) is connected to an actual terminal that will process and use them (you wouldn't want
//! these control sequences to exist in output that the user is having the terminal redirect into a
//! text file or another program as then it would just be treated as a part of the text and lead to
//! garbled output and confusion). Some utility functions are provided within the [support mod] to
//! assist you with determining appropriateness of using formatted output.
//!
//! # Predefined sequences
//!
//! There is no strict need to actually understand how to construct such control sequences to make
//! use of the formatting they offer; many predefined sequences are available for your use in the
//! [predefined mod].
//!
//! To make use of these simply inject them into the right points of a to-be-printed string via
//! format arguments, as demonstrated below. Remember not to forget to make use of a reset after the
//! text to be formatted. For instance:
//!
//! ```rust
//! use term_ctrl::predefined::{RESET, colours::fg::RED};
//! println!("{}Error:{} You made an error!", RED, RESET);
//! ```
//!
//! In the example just given, the text “<span style="color:#c00">Error:</span>” will be highlighted
//! in red. Of course as mentioned earlier, you should only use the colour sequences if output is
//! actually connected to a terminal (and that the terminal has not redirected it, so do not assume
//! output to stdout is okay). Here the example is modified to use a helper to check suitability and
//! a filter to control use:
//!
//! ```rust
//! use term_ctrl::predefined::{RESET, colours::fg::RED};
//! let format = term_ctrl::support::fmt_supported_stdout();
//! let filter = |seq| { match format { true => seq, false => "" } };
//! println!("{}Error:{} You made an error!", filter(RED), filter(RESET));
//! ```
//!
//! Note, when resetting to normal, be sure to always use the proper reset sequence. Do not make the
//! mistake of setting text colour to black and presuming that this achieves the same thing; it does
//! not. (Consider that some people have black text on a white background in their terminal, whilst
//! others use the opposite! Black text on a black background does not work very well!).
//!
//! # Constructing sequences
//!
//! Although understanding the sequences is not required, there are however benefits to be gained
//! from it, since for instance not every possible combination can reasonably be provided in
//! predefined form, and you may be able to achieve greater efficiency by constructing a custom one
//! rather than using multiple predefines in sequence.
//!
//! Let’s take a quick look at the sequence itself, then we can move on to discuss a macro which
//! helps make custom construction very easy.
//!
//! ## The sequence pattern
//!
//! The sequence pattern consists of four component parts:
//!
//! <table style="width:auto">
//!     <tr><td>1</td><td>ESC ('\u{1B}')</td><td>The escape (␛) char, Unicode 0x1B</td></tr>
//!     <tr><td>2</td><td>'['</td><td>An opening bracket char</td></tr>
//!     <tr><td>3</td><td colspan="2">One or more numbers using a semi-colon (';') as a separator</td></tr>
//!     <tr><td>4</td><td>'m'</td><td>A lower-case letter m</td></tr>
//! </table>
//!
//! In other words a pattern of: `"\u{1B}[` <em><codes\></em> `m"`
//!
//! Every control sequence fits this template, and it is the set of numbers included in it which
//! determine the precise formatting nature of the request. As a quick example, the sequence of
//! `"\u{1B}[31;1m"` specifies two numbers, `31` which corresponds to red text, and `1` which
//! corresponds to bold text (the set of available number codes are discussed shortly).
//!
//! Typically each of the individual numbers given corresponds to a particular effect, however as we
//! will see shortly there are a couple of exceptions where multiple numbers are used.
//!
//! Before we continue, understand that the effects specified are applied in sequence and remain in
//! effect until changed or reset by a subsequent code in the sequence or by such a code in a later
//! sequence. Hence in the example given earlier, a sequence is used to enable red text, this is
//! followed by some text to which this formatting applies, then a new sequence is given that issues
//! the reset-all code. There are codes available for removing specific effects or resetting just
//! foreground or background-highlight colours. There is also the catch-all code `0` for resetting
//! everything to normal.
//!
//! ## Macro construction
//!
//! As just mentioned, a macro (called [`seq`]) is provided to assist with constructing sequences.
//! All you have to do is provide it with a list of decimal numbers and it will construct a string
//! of the above pattern that contains them. An example:
//!
//! ```rust
//! use term_ctrl::seq;
//! assert_eq!("\u{1B}[1;2;3m", seq!(1, 2, 3));
//! ```
//!
//! Note that you are not restricted to pure numeric literals, string literals work also:
//!
//! ```rust
//! use term_ctrl::seq;
//! assert_eq!("\u{1B}[1;2;3m", seq!("1", "2", "3"));
//! assert_eq!("\u{1B}[1;2;3m", seq!("1;2", 3));
//! ```
//!
//! Convenience macros are also provided for constructing 256-colour and RGB colour code sets for
//! use in a sequence (they do not generate the full sequence, just a multi-code set of numbers to
//! use in a sequence).
//!
//! ## Number code chart
//!
//! The following is a guide to the available number codes and what they correspond to. In most
//! cases a single number code corresponds to a single effect, however there are also those that
//! require a sequence of multiple numbers (with the normal semi-colon separator).
//!
//! ### Effects
//!
//! <table>
//!     <thead>
//!         <tr><th>Code</th><th>Effect</th><th>Code</th><th>Effect</th></tr>
//!     </thead>
//!     <tbody>
//!         <tr><td>0</td><td>Normal (reset)</td><td>20</td><td><a href="https://en.wikipedia.org/wiki/Fraktur_(script)">Fraktur</a> on</td></tr>
//!         <tr><td>1</td><td>Bold on (increase intensity)</td><td>21</td><td>Double-underline on</td></tr>
//!         <tr><td>2</td><td>Dim on (faint, or decrease intensity)</td><td>22</td><td>Bold and dim off</td></tr>
//!         <tr><td>3</td><td>Italic on</td><td>23</td><td>Italic and Fraktur off</td></tr>
//!         <tr><td>4</td><td>Underline on</td><td>24</td><td>Underline (and double-underline) off</td></tr>
//!         <tr><td>5</td><td>Blink on</td><td>25</td><td>Blink off (aka “steady”)</td></tr>
//!         <tr><td>6</td><td>Rapid-blink on</td><td>26</td><td><em>unused?</em></td></tr>
//!         <tr><td>7</td><td>“Inverse” on (swap fg/bg colours)</td><td>27</td><td>Inverse off (aka “positive”)</td></tr>
//!         <tr><td>8</td><td>Invisible (hidden) on</td><td>28</td><td>Invisible off (aka “visible”)</td></tr>
//!         <tr><td>9</td><td>Strike-through on</td><td>29</td><td>Strike-through off</td></tr>
//!     </tbody>
//! </table>
//!
//! Code zero (`0`) resets all effects and colour selections to defaults.
//!
//! Note that codes `22`-`25` and `27`-`29` provide for removing specific effects.
//!
//! Codes `10-19` (unlisted above) are used for font selection. Code `10` selects the primary
//! (default) font. Codes `11`-`19` select alternate fonts 1-9.
//!
//! ### Basic colours
//!
//! The basic colour palettes consist of a simple set of eight colours for each of foreground and
//! background-highlight uses. There are also “bright” variants.
//!
//! <table>
//!     <thead>
//!         <tr><th>Colour</th><th>Foreground</th><th>Background</th><th>Bright foreground</th><th>Bright background</th></tr>
//!     </thead>
//!     <tbody>
//!         <tr><td>Black</td><td>30</td><td>40</td><td>90</td><td>100</td></tr>
//!         <tr><td>Red</td><td>31</td><td>41</td><td>91</td><td>101</td></tr>
//!         <tr><td>Green</td><td>32</td><td>42</td><td>92</td><td>102</td></tr>
//!         <tr><td>Yellow</td><td>33</td><td>43</td><td>93</td><td>103</td></tr>
//!         <tr><td>Blue</td><td>34</td><td>44</td><td>94</td><td>104</td></tr>
//!         <tr><td>Magenta</td><td>35</td><td>45</td><td>95</td><td>105</td></tr>
//!         <tr><td>Cyan</td><td>36</td><td>46</td><td>96</td><td>106</td></tr>
//!         <tr><td>White</td><td>37</td><td>47</td><td>97</td><td>107</td></tr>
//!         <tr><td><em>Extended</em></td><td>38</td><td>48</td><td><em>n/a</em></td><td><em>n/a</em></td></tr>
//!         <tr><td><em>Default</em> (reset)</td><td>39</td><td>49</td><td><em>n/a</em></td><td><em>n/a</em></td></tr>
//!     </tbody>
//! </table>
//!
//! The “extended” codes `38` and `48` are used to start multi-code “extended” colour palette
//! selections, as discussed shortly.
//!
//! The “default” codes `39` and `49` are used to reset foreground and background-highlight colours
//! respectively to defaults, and apply to bright and extended colouring also.
//!
//! Note that the “bright” codes are not actually a part of the official ANSI standard, apparently.
//!
//! ### Extended range colours
//!
//! Wider choice of colours than above is made available as *extended* colours through use of a
//! sequence of codes (using the normal semi-colon separator).
//!
//! There are two ranges available, an 8-bit (256 colour) range, and a 24-bit RGB range. Both begin
//! with either `38` to choose text (foreground colouring) or `48` for background-highlight
//! colouring. This is followed by `5` for specifying 256-colour or `2` for specifying RGB colour.
//! With 256-colour a final third number with a value in the range of 0-255 selects the specific
//! colour. With RGB, three numbers must be given, each corresponding to red, green and blue
//! respectively, and each also being of a value in the range 0-255.
//!
//! [See here][Xterm_256color_chart.svg] for a 256-colour chart.
//!
//! As an example, `seq!(38,5,238)` changes the colour of text (not background) since it starts with
//! `38`, it is providing a 256-colour palette selection (the `5`), and is specifically selecting
//! colour `238` from that palette.
//!
//! As another example, `seq!(48,2,180,15,70)` changes the colour of the text background since it
//! starts with `48`, it is providing an RGB colour (the `2`), and is then followed with RGB values
//! of `180` for red, `15` for green and `70` for blue.
//!
//! # Resources
//!
//! A collection of useful or informative related resources:
//!
//!  - [Wikipedia: ANSI_escape_code][wikipedia_ANSI_escape_code_SGR] (Specifically the CSI-SGR stuff)
//!  - [Wikipedia: Xterm][wikipedia_Xterm]
//!  - [Xterm control sequence documentation][xterm_ctlseqs]
//!  - [256-color chart][Xterm_256color_chart.svg]
//!
//! [support mod]: mod@crate::support
//! [predefined mod]: mod@crate::predefined
//! [`seq`]: macro@seq
//! [Xterm_256color_chart.svg]: https://upload.wikimedia.org/wikipedia/commons/1/15/Xterm_256color_chart.svg
//! [wikipedia_ANSI_escape_code_SGR]: https://en.wikipedia.org/wiki/ANSI_escape_code#SGR
//! [wikipedia_Xterm]: https://en.wikipedia.org/wiki/Xterm
//! [xterm_ctlseqs]: https://invisible-island.net/xterm/ctlseqs/ctlseqs.html

#![no_std]

extern crate atty;
#[cfg(windows)]
extern crate winapi;

pub mod codes;
mod macros;
pub mod predefined;
pub mod support;
